pkg/webhook/external/generic_actions.go

```diff
@@ -0,0 +1,95 @@
+package external
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/logging"
+)
+
+type actionResult struct {
+	ActionKind      kargoapi.GenericWebhookActionKind `json:"actionKind"`
+	ConditionResult conditionResult                   `json:"conditionResult"`
+	TargetResults   []targetResult                    `json:"targetResults,omitempty"`
+}
+
+func newActionResult(action kargoapi.GenericWebhookAction) actionResult {
+	return actionResult{
+		ActionKind: action.Kind,
+		ConditionResult: conditionResult{
+			Expression: action.MatchExpression,
+		},
+	}
+}
+
+type conditionResult struct {
+	Expression string `json:"expression"`
+	Satisfied  bool   `json:"satisfied"`
+	EvalError  string `json:"evalError,omitempty"`
+}
+
+type targetResult struct {
+	Kind           kargoapi.GenericWebhookTargetKind `json:"kind"`
+	Name           string                            `json:"name,omitempty"`
+	ListError      error                             `json:"listError,omitempty"`
+	RefreshResults []refreshResult                   `json:"refreshResults,omitempty"`
+}
+
+func newActionEnv(params map[string]string, baseEnv map[string]any) map[string]any {
+	actionEnv := maps.Clone(baseEnv)
+	m := make(map[string]any, len(params))
+	for paramKey, paramValue := range params {
+		m[paramKey] = paramValue
+	}
+	actionEnv["params"] = m
+	return actionEnv
+}
+
+func handleAction(
```

Comment:
It may be cleaner to make this a method of `genericWebhookReceiver` and then you don't need to pass these clients around. Same goes for `handleRefresh()` a few lines down.